linera-execution/src/evm/revm.rs

```diff
@@ -766,25 +595,49 @@ impl<'a> ContractPrecompile {
                 application_id,
                 stream_name,
             } => {
+                let mut runtime = context.db().0.runtime.lock().unwrap();
                 runtime.unsubscribe_from_events(chain_id, application_id, stream_name)?;
                 Ok(vec![])
             }
             ContractRuntimePrecompile::QueryService {
                 application_id,
                 query,
-            } => runtime.query_service(application_id, query),
+            } => {
+                let mut runtime = context.db().0.runtime.lock().unwrap();
+                runtime.query_service(application_id, query)
+            }
             ContractRuntimePrecompile::ValidationRound => {
+                let mut runtime = context.db().0.runtime.lock().unwrap();
                 let value = runtime.validation_round()?;
                 Ok(bcs::to_bytes(&value)?)
             }
+            ContractRuntimePrecompile::Transfer { account, amount } => {
+                if amount != Amount::ZERO {
+                    let destination = {
+                        let destination = get_evm_destination(context, account)?;
+                        destination.unwrap_or(FAUCET_ADDRESS)
+                    };
+                    let application_id = {
+                        let mut runtime = context.db().0.runtime.lock().unwrap();
+                        let application_id = runtime.application_id()?;
+                        let source = application_id.into();
+                        runtime.transfer(source, account, amount)?;
+                        application_id
+                    };
+                    let source: Address = application_id.evm_address();
+                    revm_transfer(context, source, destination, amount.into())?;
```

Comment:
Or did you mean to do this the other way round? Consider them burned if they were sent to _another_ chain? (But then what happens if they are sent to another _application_ on this chain?)

Reply:
And these are persisted as part of the EVM state, are they? But we have separate EVM states for different applications on the same chain?

What if app A sends tokens to app B on the same chain: Do we need to update both balances in both EVM states? Or are they re-initialized in each contract call, based on the actual native token balances?

Ah, that's how it has to work, does it? We never persist the balances in the EVM state?